sql/schemachanger: fix unusable secondary indexes during pk swap

fqazi opened the pull request:
Previously, the declarative schema changer during ALTER PRIMARY KEY, the declarative schema changer would make the new primary index public and then work on creating replacement secondary indexes. This could lead to broken behaviour where you could be left with either unusable old secondary indexes (causing trouble for DML queries) or potential validation errors. To address this, this patch adjusts the rules so that the new primary index is kept at a validated state, while replacement secondary indexes are constructed. This guarantees that all secondary indexes are usable during a PK swap, and the new primary key and replacement secondary indexes are swapped in an atomic fashion.

Additionally, to help support this change the index backfiller has been updated, so that a source index can be specified for backfilling indexes. This allows the index backfiller to use a validated primary index to create the secondary index, without exposing it to read queries.

Fixes: #133129
Fixes: #130165

Release note (bug fix): Addressed a bug where secondary indexes could be
unusable by DML while a primary key swap was occurring, if the new
primary key did not contain columns from the old primary key.

fqazi committed:
sql/schemachanger: fix unusable secondary indexes during pk swap

Previously, the declarative schema changer during ALTER PRIMARY KEY, the
declarative schema changer would make the new primary index public and
then work on creating replacement secondary indexes. This could lead to
broken behaviour where you could be left with either unusable old
secondary indexes (causing trouble for DML queries) or potential
validation errors. To address this, this patch adjusts the rules so that
the new primary index is kept at a validated state, while replacement
secondary indexes are constructed. This guarantees that all secondary
indexes are usable during a PK swap, and the new primary key and
replacement secondary indexes are swapped in an atomic fashion.

Fixes: #133129
Fixes: #130165

Release note (bug fix): Addressed a bug where secondary indexes could be
unusable by DML while a primary key swap was occurring, if the new
primary key did not contain columns from the old primary key.

fqazi committed:
sql: backfill indexes using sourceIndexID

Previously, for the index backfiller only used the current primary index
as the source of backfills. This works correctly in all cases today,
because we only backfill from public indexes. However, for primary key
changes we need the ability to backfill secondary indexes from validate
primary indexes. This would allow us to support transactional behavior
and avoid edge cases where secondary indexes are not usable today. To
address this, this patch introduces a source index ID into the
backfiller that when populated will be used instead of the primary
index.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141850)
<!-- Reviewable:end -->

spilchen commented:
I haven’t reviewed all the changes yet (I will), but one early comment is that a few tests were disabled that I believe can now be re-enabled. Search for 133129 in:
- pkg/ccl/testccl/sqlccl/explain_test.go
- pkg/sql/schemachanger/dml_injection_test.go

fqazi commented:
@spilchen @yuzefovich Done, and thank you for tracking that down! Think we only need a foundations review for this one, since the backfiller changes are pretty trivial.

fqazi commented:
@spilchen RFAL

spilchen reviewed (commented):
Very nice change! This is much simpler than my early attempt at this problem.

> Reviewed 3 of 7 files at r1, 15 of 126 files at r7, 19 of 126 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141850#-:-OL5TdACBk2ngqzWwLDe:b-nl7q9x)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @golgeek, @herkolategan, and @jeffswenson)

___
*[`pkg/sql/execinfrapb/processors_bulk_io.proto` line 86 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/141850#-OL54AlI4tKZ1PVyADVw:-OL54AlI4tKZ1PVyADVx:b5xzz0m) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ba901a4e3dbe5e7a28c4ac096ae4543b86ebf61/pkg/sql/execinfrapb/processors_bulk_io.proto#L86)):*
> ```proto
>   // SourceIndexID is the source index that will be used for this backfill. If
>   // the value of 0 is specified then the primary index will be used implicitly.
>   optional uint32 source_index_id = 16 [(gogoproto.nullable) = false, (gogoproto.customname) = "SourceIndexID", (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb.IndexID"];
> ```

nit: should this be field number 15? Or, should the NEXTID comment below have 17?
___
*[`pkg/sql/catalog/tabledesc/validate.go` line 1794 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/141850#-OL57RNnAbqVlbLwOhIP:-OL57RNnAbqVlbLwOhIQ:b7amfke) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0e6116cc597eb37f4852266c083fb5a3648f589/pkg/sql/catalog/tabledesc/validate.go#L1794)):*
> ```Go
> 
> 			// When newPKColIDs is not empty, it means there is an in-progress `ALTER
> 			// PRIMARY KEY`. Certain schema changes will make the new virtual columns
> ```

Is there any concern that, before the new PK becomes public, the plan might use the index under the assumption that the virtual column will be included?
___
*[`pkg/sql/backfill/backfill.go` line 767 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/141850#-OL53_2P2JwTMQitsxxd:-OL53_2P2JwTMQitsxxe:b-t2axma) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ba901a4e3dbe5e7a28c4ac096ae4543b86ebf61/pkg/sql/backfill/backfill.go#L767)):*
> ```Go
> 	if sourceIndexID != 0 {
> 		ib.sourceIndex = catalog.FindIndexByID(desc, sourceIndexID)
> 	}
> ```

nit: is there a reason why need to call `desc.GetPrimaryIndex` if we aren't intending to use it? Could we restructure the code differently.



_Suggestion:_
```Go
	if sourceIndexID != 0 {
		ib.sourceIndex = catalog.FindIndexByID(desc, sourceIndexID)
	} else {
		ib.sourceIndex = desc.GetPrimaryIndex()
	}
```


<!-- Sent from Reviewable.io -->

fqazi committed:
sql/schemachanger: prevent use of recreated secondary indexes

Previously, when secondary indexes were recreated during a primary index
change, they were made public before the new primary index was fully
public. This posed a risk: queries could potentially access the
secondary indexes before they were fully ready. To address this, this
patch makes secondary index replacements invisible until the new primary
index backing them is ready.

Release note: None

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141850)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @golgeek, @herkolategan, @jeffswenson, and @spilchen)

___
*[`pkg/sql/catalog/tabledesc/validate.go` line 1794 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/141850#-OL57RNnAbqVlbLwOhIP:-OLetUf72A8xCnz3rN8N:b-twtpjk) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0e6116cc597eb37f4852266c083fb5a3648f589/pkg/sql/catalog/tabledesc/validate.go#L1794)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Is there any concern that, before the new PK becomes public, the plan might use the index under the assumption that the virtual column will be included?
</blockquote></details>

Done.


Fair point, and I added another commit so that the new secondary indexes are temporarily invisible, until the primary index is ready. This prevents us from using them to service queries.

I tried using rules to order things, but it would be cyclic if your dropping a column that is referenced in a secondary index.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
RFAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141850#-:-OLeuHWD5i_PCRu2SZWz:bb72pd0)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @golgeek, @herkolategan, @jeffswenson, and @spilchen)



<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> Reviewed 127 of 127 files at r9, 126 of 126 files at r10, 13 of 13 files at r12, 17 of 17 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141850)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @golgeek, @herkolategan, and @jeffswenson)

___
*[`pkg/sql/schemachanger/scplan/internal/opgen/opgen_primary_index.go` line 130 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/141850#-OLire6H8I--iO-ybqzd:-OLire6H8I--iO-ybqze:b-rm7013) ([raw file](https://github.com/cockroachdb/cockroach/blob/0888a1ea1620ded195ed89a38bd8dc286d9329ce/pkg/sql/schemachanger/scplan/internal/opgen/opgen_primary_index.go#L130)):*
> ```Go
> 						}
> 						// Index was created as invisible, nothing needs to be done.
> 						if idx.IsNotVisible || idx.Invisibility == 1.0 {
> ```

We have already checked `idx.IsNotVisible` a few lines above. Was the intent here to check the visibility of the primary index? i.e. `this.IsNotVisible || this.Invisibility == 1.0`. If not, can we merge the if check.
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 346 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/141850#-OLir8SO23sNulw1JXL7:-OLir8SO23sNulw1JXL8:b-fgjj2q) ([raw file](https://github.com/cockroachdb/cockroach/blob/0888a1ea1620ded195ed89a38bd8dc286d9329ce/pkg/sql/schemachanger/scpb/elements.proto#L346)):*
> ```proto
>   uint32 recreate_source_id = 4 [(gogoproto.customname) = "RecreateSourceIndexID", (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/sem/catid.IndexID"];
>   // If an index is being recreated, this is the target primary
>   // index that will make it visible for usage.
> ```

Can we clarify that this is the ID of the new primary index? There is also `SourceIndexID`, which based on the discussion yesterday on slack (https://cockroachlabs.slack.com/archives/C04N0AS14CT/p1742314411019249), is the ID of the primary index too, but the old (?) one.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141850)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @golgeek, @herkolategan, @jeffswenson, and @spilchen)

___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 346 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/141850#-OLir8SO23sNulw1JXL7:-OLiuGKu-kSB6bFepQWI:bbt476n) ([raw file](https://github.com/cockroachdb/cockroach/blob/0888a1ea1620ded195ed89a38bd8dc286d9329ce/pkg/sql/schemachanger/scpb/elements.proto#L346)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Can we clarify that this is the ID of the new primary index? There is also `SourceIndexID`, which based on the discussion yesterday on slack (https://cockroachlabs.slack.com/archives/C04N0AS14CT/p1742314411019249), is the ID of the primary index too, but the old (?) one.
</blockquote></details>

Done.

I made it clear this is the final index that needs to be public:

// If an index is being recreated, this is the final primary  
// index that will make it usable (i.e. the columns required  
// to back this index will be public in this index).
___
*[`pkg/sql/schemachanger/scplan/internal/opgen/opgen_primary_index.go` line 130 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/141850#-OLire6H8I--iO-ybqzd:-OLiuCotAQwmwkygXwB-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0888a1ea1620ded195ed89a38bd8dc286d9329ce/pkg/sql/schemachanger/scplan/internal/opgen/opgen_primary_index.go#L130)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

We have already checked `idx.IsNotVisible` a few lines above. Was the intent here to check the visibility of the primary index? i.e. `this.IsNotVisible || this.Invisibility == 1.0`. If not, can we merge the if check.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

spilchen reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 17 of 17 files at r14, 17 of 17 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141850#-:-OLjB6kI7ySsWB7M9MJz:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @andyyang890, @golgeek, @herkolategan, and @jeffswenson)

___
*[`pkg/sql/schemachanger/scexec/scmutationexec/index.go` line 520 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/141850#-OLjAArS9qCFcbrVf6NY:-OLjAArS9qCFcbrVf6NZ:b6zzob8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d1aac88b3adcbbbff29cad2714ba2044d953899/pkg/sql/schemachanger/scexec/scmutationexec/index.go#L520)):*
> ```Go
> 	// If the primary index is already visible, then nothing
> 	// needs to be done. This can happen if we are able to sequence
> 	// publishing the primary and secondary indexes together.g
> ```

nit: spelling



_Suggestion:_
```Go
// publishing the primary and secondary indexes together.
```
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 347 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/141850#-OLjASh6AYMwG3yY3SZI:-OLjASh6AYMwG3yY3SZJ:bdapdr2) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d1aac88b3adcbbbff29cad2714ba2044d953899/pkg/sql/schemachanger/scpb/elements.proto#L347)):*
> ```proto
>   // If an index is being recreated, this is the final primary
>   // index that will make it usable (i.e. the columns required
>   // to back this index be will public in this index).
> ```

nit: suggest slight rewording



_Suggestion:_
```proto
// to back this index will be public in this index)
```


<!-- Sent from Reviewable.io -->

fqazi commented:
@spilchen TFTR!

bors r+
